### PR TITLE
Remove hir::Ident

### PR DESCRIPTION
**Pull request by petrochenkov**
Now when name resolution is done on AST, `hir::Ident` is no longer necessary.
See https://github.com/rust-lang/rust/pull/30145 for more details.

r? @nrc

### CODE REVIEW

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -862,9 +865,14 @@ impl<'a> LoweringContext<'a> {
                 PatKind::Wild => hir::PatKind::Wild,
                 PatKind::Ident(ref binding_mode, pth1, ref sub) => {
                     self.with_parent_def(p.id, |this| {
+                        let name = match this.resolver.get_resolution(p.id).map(|d| d.full_def()) {
+                            // Only pattern bindings are renamed
```
Comment (nrc) [inline src/librustc/hir/lowering.rs]: Why do we only sometimes rename here?

Comment (petrochenkov) [inline src/librustc/hir/lowering.rs]: We don't want to rename constants or unit variants/structs, that can also be represented with `PatKind::Ident` due to parsing ambiguities, only fresh bindings.

Comment (nrc) [inline src/librustc/hir/lowering.rs]: At this stage we are post-parsing and post-name resolution, so where do the remaining ambiguities get resolved? Put another way, in some sense this renaming must happen somewhere, where does it happen (and how)?

Comment (petrochenkov) [inline src/librustc/hir/lowering.rs]: I'm not sure I understand the question.
>At this stage we are post-parsing and post-name resolution, so where do the remaining ambiguities get resolved?

All ambiguities are resolved during name resolution, there's nothing ambiguous at this stage.

>Put another way, in some sense this renaming must happen somewhere, where does it happen (and how)?

Renaming is done twice, once during name resolution, its results aren't saved anywhere, and once more during lowering to HIR, its results are persistent and kept in HIR.
`Name`s in HIR need to be properly renamed to correctly treat cases like [this](https://github.com/rust-lang/rust/pull/29748#issuecomment-155532904), which in principle can be interpreted as requiring some rudimentary form of name resolution during type checking.

Comment (petrochenkov) [inline src/librustc/hir/lowering.rs]: I tables of "canonical bindings" were filled during the main name resolution pass and saved for later, then HIR could keep plain non-renamed `Name`s for everything.

**Comment by petrochenkov** [comment]
Rebased.

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (nrc) [inline src/librustc/hir/lowering.rs]: thanks for the explanation!

**Pull request merged**